### ISSUE TITLE
'named symbol not found' error at runtime in cuda_piKernelCreate if static library with kernel is linked

### ISSUE THREAD

**Issue by ivorobts** [intel/llvm#4307]
A simple example showing the usage of static library with kernel which fails at runtime with CUDA BE:
Here is the content of file sycl_lib.cpp:

```
#include<CL/sycl.hpp>
using namespace sycl;
void func() {
  queue q;
  q.submit([&](sycl::handler &h) {
      sycl::stream os(1024, 768, h);
      h.parallel_for(32, [=](sycl::id<1> i) {
          os<<i<<"\n";
        });
    });
}
```

and a simple main code:
```
void func();
int main() {
  func();
  return 0;
}
```

Compilation of sycl_lib.cpp, static library creation and linking works fine:

```
clang++  -fsycl -fsycl-targets=nvptx64-nvidia-cuda-sycldevice -fsycl-unnamed-lambda -c sycl_lib.cpp
ar rvs sycl_lib.a sycl_lib.o
clang++ -fsycl -fsycl-targets=nvptx64-nvidia-cuda-sycldevice main.cpp -o test-lib sycl_lib.a
SYCL_BE=PI_CUDA ./test-lib
```

The application crashes at runtime with following error:

```
PI CUDA ERROR:
        Value:           500
        Name:            CUDA_ERROR_NOT_FOUND
        Description:     named symbol not found
        Function:        cuda_piKernelCreate
        Source Location: /llvm/sycl/plugins/cuda/pi_cuda.cpp:2380

terminate called after throwing an instance of 'cl::sycl::runtime_error'
  what():  Native API failed. Native API returns: -999 (Unknown OpenCL error code) -999 (Unknown OpenCL error code)
```

Note that ahead of time compilation and linking works fine for CPU or Intel GPU devices. This looks like a limitation of CUDA BE.

**Comment by Michoumichmich** [comment]
Hello, I get also the same issue on the CUDA BE, but this is not a limitation of the CUDA back-end. Could you try: https://github.com/Michoumichmich/llvm/commit/e3eb050b73bff24cb8335f84fa6a51f1ebaf05c5.diff it is a bug that was introduced by f7ce532d622dc26eddd25f87faec0ff35dc0c2e9 in a pulldown

**Comment by ravil-mobile** [comment]
Thanks a lot. It saved my day.

**Comment by Michoumichmich** [comment]
> Thanks a lot. It saved my day.
You're welcome!

**Comment by danchitnis** [comment]
It seems to happen on some versions but not others. It happens both on WSL and native Linux, I assumed it was related to GPU driver mismatch with CUDA. perhaps can you make the error info more useful with debugging traces?

**Comment by Michoumichmich** [comment]
> It seems to happen on some versions but not others. It happens both on WSL and native Linux, I assumed it was related to GPU driver mismatch with CUDA. perhaps can you make the error info more useful with debugging traces?

The error is coming from the fact that the names of the generated device object files were changed, but `clang-offload-bundler` is still called on the old names it fails silently. If you compile with `-v` you won't even see `ptxas` compiling down the kernels during the final linking of the `.so` object or executable

**Comment by danchitnis** [comment]
So far it is working for me, but I am not using the latest commit (a month old). I am aware that there is a recent bug with nvidia containers influencing WSL (and native Linux), so then these two bugs may not be related.

**Comment by Michoumichmich** [comment]
Hello @ivorobts can you try compiling with `nvptx64-nvidia-cuda--sm_75`  on the current branch to see if it's fixed ?

**Comment by npmiller** [comment]
This was fixed recently by this PR:
* https://github.com/intel/llvm/pull/4305

If you want to try it note that your SYCL code is missing a `wait` which might cause issues at the moment, something like this works fine though:
```cpp
#include<CL/sycl.hpp>
using namespace sycl;
void func() {
  queue q;
  q.submit([&](sycl::handler &h) {
      sycl::stream os(1024, 768, h);
      h.parallel_for(32, [=](sycl::id<1> i) {
          os<<i<<"\n";
        });
    }).wait_and_throw();
}
```

I'm not entirely sure if it should work without the `wait` but either way that would be a separate issue to investigate, the original problem reported here should be fixed.

**Comment by npmiller** [comment]
Just as a side note, I double checked with the SYCL 2020 specification and the `wait` is indeed required in this case, the queue destructor doesn't not implicitly wait on kernel completion, section 3.9.8.1:

> Note that the destructors of other SYCL objects (sycl::queue, sycl::context,...) do not block. Only a
sycl::buffer, sycl::sampled_image or sycl::unsampled_image destructor might block. The rationale is that
an object without any side effect on the host does not need to block on destruction as it would impact the
performance. So it is up to the programmer to use a member function to wait for completion in some
cases if this does not fit the goal.

**Comment by AerialMantis** [comment]
We believe this issue to be resolved now, so we are closing the ticket, @ivorobts if you continue to see any issue here please feel free to comment and we can re-open the ticket.

**Comment by Luigi-Crisci** [comment]
Hi,
I'm having the same issue with the lastest build from the sycl branch when I try to compile the same code provided by @ivorobts. This happens when targeting `nvptx64_nvidia_cuda`.

> This was fixed recently by this PR:
> 
> * [[Driver] Properly set offload-deps target triple #4305](https://github.com/intel/llvm/pull/4305)
> 
> If you want to try it note that your SYCL code is missing a `wait` which might cause issues at the moment, something like this works fine though:
> 
> ```c++
> #include<CL/sycl.hpp>
> using namespace sycl;
> void func() {
>   queue q;
>   q.submit([&](sycl::handler &h) {
>       sycl::stream os(1024, 768, h);
>       h.parallel_for(32, [=](sycl::id<1> i) {
>           os<<i<<"\n";
>         });
>     }).wait_and_throw();
> }
> ```
> 
> I'm not entirely sure if it should work without the `wait` but either way that would be a separate issue to investigate, the original problem reported here should be fixed.

I see that it should have been fixed, maybe the lastest build broke it again?

**Comment by bader** [comment]
Re-open the issue to check again.

**Comment by npmiller** [comment]
I had a look and it is indeed failing again with the most recent build, I was able to track it down to the following commit:
* https://github.com/intel/llvm/commit/98380762c3b734c23d206182605ab9e035c93caa

I'll look into fixing this again.
